File: chatapp-mobile/app/src/main/java/com/chatapp/mobile/models/Message.java
package com.chatapp.mobile.models;

import java.util.List;

public class Message {
    public String _id;
    public Object sender; // Can be a User object or a String ID depending on query
    public String receiver;
    public String messageType; // "text", "image", "video"
    public String text;
    public String fileUrl;
    public String status; // "sent", "delivered", "read"
    public String createdAt;
    public boolean isDeletedForEveryone;
    public List<Reaction> reactions;

    public static class Reaction {
        public String user;
        public String emoji;
    }

    public String getSenderId() {
        if (sender instanceof String) {
            return (String) sender;
        } else if (sender instanceof com.google.gson.internal.LinkedTreeMap) {
            // Gson can parse nested objects as LinkedTreeMap
            com.google.gson.internal.LinkedTreeMap map = (com.google.gson.internal.LinkedTreeMap) sender;
            return (String) map.get("_id");
        } else if (sender instanceof User) {
            return ((User) sender)._id;
        }
        return "";
    }
}

File: server/models/Message.js
import mongoose from "mongoose";

/*
  Message Status Flow (like WhatsApp):
  ─────────────────────────────────────
  "sent"      → Server received the message      (single grey tick  ✓ )
  "delivered" → Recipient's device received it   (double grey tick  ✓✓)
  "read"      → Recipient opened the chat        (double blue tick  ✓✓ in blue)
*/

const messageSchema = new mongoose.Schema(
  {
    // ── Participants ───────────────────────────────────────
    sender: {
      type: mongoose.Schema.Types.ObjectId,
      ref: "User",
      required: true,
    },

    receiver: {
      type: mongoose.Schema.Types.ObjectId,
      ref: "User",
      required: true,
    },

    // ── Content ────────────────────────────────────────────
    messageType: {
      type: String,
      enum: ["text", "image", "video"],
      default: "text",
    },

    text: {
      type: String,
      trim: true,
      maxlength: [5000, "Message cannot exceed 5000 characters"],
      default: "",
    },

    fileUrl: {
      type: String,
      default: "",
    },

    // ── Message Reactions ──────────────────────────────────
    reactions: [
      {
        user: {
          type: mongoose.Schema.Types.ObjectId,
          ref: "User",
          required: true,
        },
        emoji: {
          type: String,
          required: true,
        },
      },
    ],

    // ── Delivery Status ────────────────────────────────────
    status: {
      type: String,
      enum: ["sent", "delivered", "read"],
      default: "sent",
    },

    // ── Soft Delete (like WhatsApp "delete for me") ────────
    deletedBy: [
      {
        type: mongoose.Schema.Types.ObjectId,
        ref: "User",
      },
    ],

    isDeletedForEveryone: {
      type: Boolean,
      default: false,
    },
  },
  {
    timestamps: true,
  }
);

// ── Index for fast conversation queries ───────────────────
messageSchema.index({ sender: 1, receiver: 1 });
messageSchema.index({ createdAt: -1 });

const Message = mongoose.model("Message", messageSchema);

export default Message;